mmda/predictors/hf_predictors/bibentry_predictor.py

```diff
@@ -0,0 +1,217 @@
+from enum import Enum
+import re
+from typing import Dict, List, Optional
+
+from pydantic import BaseModel
+from transformers import AutoConfig, AutoTokenizer, AutoModelForTokenClassification
+from unidecode import unidecode
+
+
+class Label(Enum):
+    MISC = 0
+
+    CITATION_NUMBER = 1
+
+    AUTHOR_START = 2
+    AUTHOR_MIDDLE = 3
+    AUTHOR_END = 4
+
+    ISSUED_DAY = 5
+    ISSUED_MONTH = 6
+    ISSUED_YEAR = 7
+
+    TITLE = 8
+    JOURNAL = 9
+    PUBLISHER = 10
+    VOLUME = 11
+    ISSUE = 12
+    PAGE = 13
+    URL = 14
+    DOI = 15
+    EVENT = 16
+    ISBN = 17
+
+
+class StringWithSpan(BaseModel):
+    content: str
+    start: int  # inclusive
+    end: int  # exclusive
+
+
+class PredictionWithSpan(BaseModel):
+    citation_number: Optional[StringWithSpan]
+    authors: Optional[List[StringWithSpan]]
+    title: Optional[StringWithSpan]
+    journal_venue_or_event: Optional[StringWithSpan]
+    year: Optional[StringWithSpan]
+    doi: Optional[StringWithSpan]
+    url: Optional[StringWithSpan]
+
+
+class BibEntryPredictor:
+    def __init__(self, model_name_or_path: str):
+        self.model = AutoModelForTokenClassification.from_pretrained(model_name_or_path)
+        self.config = AutoConfig.from_pretrained(model_name_or_path)
+        self.tokenizer = AutoTokenizer.from_pretrained(model_name_or_path)    
+    
+    def predict(self, bib_entries: List[str]) -> List[PredictionWithSpan]:
+        # TODO: eventually, this input will be a Document
+        # and this function should have argstring: def predict(self, document: Document) -> List[Annotation]:
+        # and some processing that looks like:
+        # for bib_entry in document.bib_entries:
+        #     bib_entry.text -> treat this as text
+        res = []
+
+        tokenized_inputs = self.tokenizer(bib_entries, padding=True, truncation=True, return_tensors="pt")
+        predictions = self.model(**tokenized_inputs)
+
+        pred_ids = predictions.logits.argmax(2).tolist()  # check GPU vs CPU; might need to .detach()
```

Comment:
Kyle mentioned to `check GPU vs CPU, might need .detach()`. I checked that it works fine on both.